### PR TITLE
Add documentation for the DYN control flow block

### PR DESCRIPTION
**Pull request by grjte**
This PR adds the new DYN control flow block to the operation and program docs, including defining the decoder constraints. This documentation matches the implementation introduced in #1055. It does not address open questions such as how to restrict `DYN` blocks to only the root context.

### CODE REVIEW

File: docs/src/design/decoder/constraints.md [hunk 1]
```diff
@@ -18,15 +18,16 @@ AIR constraints for the decoder involve operations listed in the table below. Fo
 | `REPEAT`  | $f_{repeat}$  | 4      | Top stack element is dropped.                                                                    |
 | `SPAN`    | $f_{span}$    | 5      | Stack remains unchanged.                                                                         |
 | `RESPAN`  | $f_{respan}$  | 4      | Stack remains unchanged.                                                                         |
+| `DYN`     | $f_{dyn}$     | 5      | Stack remains unchanged.                                                                         |
 | `CALL`    | $f_{call}$    | 4      | Top stack element is dropped.                                                                    |
 | `SYSCALL` | $f_{syscall}$ | 4      | Top stack element is dropped.                                                                    |
```
Comment (bobbinth) [inline docs/src/design/decoder/constraints.md:23]: Not related to this PR, but I think there might be an error here for `CALL` and `SYSCALL` as I don't think we drop the top stack element for this operations. Could you check this?

File: docs/src/design/decoder/constraints.md [hunk 2]
```diff
@@ -38,6 +39,12 @@ When `SPLIT` or `LOOP` operation is executed, the top of the operand stack must
 (f_{split} + f_{loop}) \cdot (s_0^2 - s_0) = 0 \text{ | degree} = 7
 $$
 
+When a `DYN` operation is executed, the hasher registers must all be set to $0$:
+
+> $$
+f_{dyn} \cdot (1 - h_i) = 0 \text { for } i \in [0, 8) \text{ | degree} = 2
+$$
```
Comment (bobbinth) [inline docs/src/design/decoder/constraints.md:46]: Should this not be degree $6$ since degree of $f_{dyn}$ is $5$?

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a couple of minor comments inline. Once these are addressed, we can merge.

File: docs/src/design/stack/op_constraints.md [hunk 3]
```diff
@@ -292,5 +292,5 @@ f_{end,repeat,respan,halt} = e_1 \cdot b_4  \text{ | degree} = 2
 $$
 
 $$
-f_{ctrl} = f_{span,join,split,loop} + f_{end,repeat,respan,halt} + f_{call} + f_{syscall} \text{ | degree} = 3
+f_{ctrl} = f_{span,join,split,loop} + f_{end,repeat,respan,halt} + f_{dyn} + f_{call} + f_{syscall} \text{ | degree} = 5
```
[the remark this answers is not in the document]
Comment (grjte) [reply docs/src/design/stack/op_constraints.md:295]: It doesn't seem to - I did check on this, but the only thing I could find that actually uses this flag is the constraint for disabling/forcing control flow operations based on the `in_span` flag. The degree of that constraint was the same as the degree of this flag (and I updated it in these docs changes), so it didn't seem to be a problem

File: docs/src/design/decoder/constraints.md (hunk 2 above)
Comment (grjte) [reply docs/src/design/decoder/constraints.md:46]: Yes! That's definitely a typo. I'll fix it in a follow-up PR

File: docs/src/design/decoder/constraints.md (hunk 1 above)
Comment (grjte) [reply docs/src/design/decoder/constraints.md:23]: Yeah, they seem to leave the stack unchanged. I'll update it in a separate PR

**Pull request merged**